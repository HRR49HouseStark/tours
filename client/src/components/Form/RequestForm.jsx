import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

import { validate } from '../../utils/Logic';

import FinancingCheck from '../FinancingCheck';
import FormErrors from './FormErrors';

// TO DO: Need property name from whoever's module that is.
const RequestForm = ({ tour, call, setCall, submit }) => {
  const [form, setForm] = useState({
    name: '', number: '', email: '', message: '',
  });
  const [errors, setError] = useState({ name: false, number: false, email: false });

  useEffect(() => {
    setForm((state) => ({ ...state, message: 'I am interested in this property!' }));
  }, []);

  const verify = (e) => validate(e.target.name, form, setError);

  const verifyAll = () => {
    ['name', 'number', 'email'].forEach((name) => validate(name, form, setError));
    return !!(errors.name || errors.number || errors.email);
  };

  const onChange = (e) => {
    setForm((state) => ({ ...state, [e.target.name]: e.target.value }));
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    const toSubmit = { ...form };
    tour ? delete toSubmit.message : null;
    verifyAll ? submit(toSubmit) : null;
  };

  const phoneRegex = '^\\(?([0-9]{3})\\)?[-.●]?([0-9]{3})[-.●]?([0-9]{4})$';
  const buttonName = tour ? 'Schedule A Tour' : 'Request Info';
  const interested = (<textarea id="reqMessage" name="message" value={form.message} onChange={onChange} className="inputField" />);

  return (
    <form id="userForm" onSubmit={handleSubmit}>

      <div id="namePhoneFlex">
        <input name="name" value={form.name} placeholder="Name" onChange={onChange} onBlur={verify} className="inputField" required />
        <input name="number" type="tel" pattern={phoneRegex} value={form.number} placeholder="Phone" onChange={onChange} onBlur={verify} className="inputField" required />
      </div>
      <FormErrors name={errors.name} number={errors.number} />

      <input name="email" id="emailForm" type="email" value={form.email} placeholder="Email" onChange={onChange} onBlur={verify} className="inputField" required />
      <FormErrors email={errors.email} />

      {tour ? (<FinancingCheck call={call} setCall={setCall} />) : interested}

      <button id="tourSubmit" type="submit">{buttonName}</button>
      {!tour ? (<FinancingCheck call={call} setCall={setCall} />) : null}
    </form>
  );
};

RequestForm.propTypes = {
  tour: PropTypes.bool.isRequired,
  call: PropTypes.bool.isRequired,
  setCall: PropTypes.func.isRequired,
  submit: PropTypes.func.isRequired,
};

export default RequestForm;
