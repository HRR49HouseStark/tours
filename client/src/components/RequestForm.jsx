import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

import { validate } from '../utils/Logic';

import FinancingCheck from './FinancingCheck';
import FormErrors from './FormErrors';

/*
TODO:
Need property name from whoever's module that is.
Prevent submit if any errors
*/

const RequestForm = ({ tour, call, setCall, submit }) => {
  const [form, setForm] = useState({
    name: '', number: '', email: '', message: '',
  });
  const [errors, setError] = useState({ name: false, number: false, email: false });

  useEffect(() => {
    setForm((state) => ({ ...state, message: 'I am interested in this property!' }));
  }, []);

  const verify = (e) => validate(e.target.name, form, setError);

  const onChange = (e) => {
    setForm((state) => ({ ...state, [e.target.name]: e.target.value }));
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    submit(form);
  };

  const phoneRegex = '^\\(?([0-9]{3})\\)?[-.●]?([0-9]{3})[-.●]?([0-9]{4})$';
  const buttonName = tour ? 'Schedule A Tour' : 'Request Info';
  const interested = (<textarea name="message" value={form.message} onChange={onChange} className="inputField" />);

  return (
    <form onSubmit={handleSubmit}>

      <div id="namePhoneFlex">
        <input name="name" value={form.name} placeholder="Name" onChange={onChange} onBlur={verify} className="inputField" required />
        <input name="number" type="tel" pattern={phoneRegex} value={form.number} placeholder="Phone" onChange={onChange} onBlur={verify} className="inputField" required />
      </div>
      <FormErrors name={errors.name} number={errors.number} />

      <input name="email" id="emailForm" type="email" value={form.email} placeholder="Email" onChange={onChange} onBlur={verify} className="inputField" required />
      <FormErrors email={errors.email} />

      {tour ? (<FinancingCheck call={call} setCall={setCall} />) : interested}

      <button type="submit">{buttonName}</button>
    </form>
  );
};

RequestForm.propTypes = {
  tour: PropTypes.bool.isRequired,
  call: PropTypes.bool.isRequired,
  setCall: PropTypes.func.isRequired,
  submit: PropTypes.func.isRequired,
};

export default RequestForm;
