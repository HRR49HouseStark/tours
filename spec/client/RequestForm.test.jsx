import React from 'react';
import { shallow } from 'enzyme';

import RequestForm from '../../client/src/components/Form/RequestForm';

describe('RequestForm Test Suite', () => {
  const a = () => {};
  const b = () => {};
  const wrapper = shallow(<RequestForm tour call setCall={a} submit={b} />);

  test('Should render', () => {
    expect(wrapper.exists()).toBeTruthy();
  });

  test('snapshot render should match', () => {
    const spy = jest.spyOn(console, 'error').mockImplementation(() => {});
    expect(wrapper).toMatchSnapshot();
    spy.mockRestore();
  });
});
