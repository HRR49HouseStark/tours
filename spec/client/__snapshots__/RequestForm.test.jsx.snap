// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestForm Test Suite snapshot render should match 1`] = `
<form
  id="userForm"
  onSubmit={[Function]}
>
  <div
    id="namePhoneFlex"
  >
    <input
      className="inputField"
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Name"
      required={true}
      value=""
    />
    <input
      className="inputField"
      name="number"
      onBlur={[Function]}
      onChange={[Function]}
      pattern="^\\\\(?([0-9]{3})\\\\)?[-.●]?([0-9]{3})[-.●]?([0-9]{4})$"
      placeholder="Phone"
      required={true}
      type="tel"
      value=""
    />
  </div>
  <FormErrors
    email={null}
    name={false}
    number={false}
  />
  <input
    className="inputField"
    id="emailForm"
    name="email"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Email"
    required={true}
    type="email"
    value=""
  />
  <FormErrors
    email={false}
    name={null}
    number={null}
  />
  <FinancingCheck
    call={true}
    setCall={[Function]}
  />
  <button
    id="tourSubmit"
    type="submit"
  >
    Schedule A Tour
  </button>
</form>
`;
